.gas-snapshot

```diff
@@ -0,0 +1,4 @@
+OCR2DROracle_setDONPublicKey:testEmptyPublicKeyReverts() (gas: 10936)
+OCR2DROracle_setDONPublicKey:testOnlyOwnerReverts() (gas: 11327)
+OCR2DROracle_setDONPublicKey:testSetDONPublicKeySuccess() (gas: 126336)
```

Comment:
IMO this value is misleading, we should ask ourselves, "What do we actually want this file to track?". At the moment, this line is tracking the GAS cost of all these lines:

```
        bytes memory donPublicKey = abi.encode("newKey");

        // Verify the existing key is different from the new key
        bytes memory existingDonPublicKey = s_oracle.getDONPublicKey();
        bytes memory expectedExistingDonPublicKey;
        assertEq(existingDonPublicKey, expectedExistingDonPublicKey);
        // If they have different lengths they are not the same
        assertFalse(donPublicKey.length == expectedExistingDonPublicKey.length);

        s_oracle.setDONPublicKey(donPublicKey);
        bytes memory newDonPublicKey = s_oracle.getDONPublicKey();
        assertEq(newDonPublicKey, donPublicKey);
```

When really, all we want to measure is the successful call of this:

```
s_oracle.setDONPublicKey(donPublicKey)
```

Whilst the current test is a valid test, it's not a valid GAS measurement for this function because it includes a bunch of things that aren't strictly what we want to measure (assertions, calling other functions and contracts before/after the target function call). Whilst there is value in these holistic measurements, from a developers perspective I think this could end up becoming noise and of little actual use.

We should instead separate out GAS specific measurements that only measure the function call we're actually targeting, and only track those values. Otherwise, every single test would be printed here and any single change to a distantly related contract would change this file in potentially hundreds/thousands of lines, hiding the actual gas effect of a change.

We've implemented this internally in another project by having test contracts separate to the usual unit tests, for example, `OCR2DROracle_Gas`, which sets up the correct state before each function, and then just runs that single function without assertions or any other lines. The gas measurement printed here is then the exact gas cost of calling that function. @DeividasK @cds95 @yosriady have some experience working this way.

We can use [filters](https://book.getfoundry.sh/forge/gas-snapshots#filtering) to produce and check the snapshot file here that only tracks these pure GAS values. Any change to this file can be analysed in a meaningful way, because the gas measurements represent the target function call and only the target function call.

Reply:
Hmm I do agree that it would be quite noisy to have all functions be documented. I do propose another solution: suffix all tests you'd like to have included into the snapshot with `_gas`. When we run `forge snapshot --match-test _gas` this should only include those functions in the snapshot. This keeps all the tests in the same place.